### PR TITLE
Vr controller integration

### LINKED ISSUE #391: Implement OpenVR (Controls)

### PR DESCRIPTION
**Pull request by AncientFungus**
Fixes #391

### CODE REVIEW

**Comment by NinjaDanz3r** [comment]
Can one of the admins verify this patch?

File: Game/Resources.json.orig [hunk 1]
```diff
@@ -0,0 +1,115 @@
+{
+<<<<<<< HEAD
```
Comment (LarsIngo) [inline Game/Resources.json.orig]: Remove file?

File: Game/Resources/Lever.as [hunk 2]
```diff
@@ -15,13 +15,13 @@ class Lever {
     
     void Update(float deltaTime){
         if (Input(MouseOpen)){
-            if(IsIntersect(self, player)){
+            //if(IsIntersect(self, player)){
```
Comment (LarsIngo) [inline Game/Resources/Lever.as]: Should be used or removed?

File: Game/Resources/Player.as [hunk 3]
```diff
@@ -10,11 +10,11 @@ class Player{
     }
     
     void Update(float i){
-        if(Cursor.x == 0 && Cursor.y == 0)
+        /*if(Cursor.x == 0 && Cursor.y == 0)
```
Comment (LarsIngo) [inline Game/Resources/Player.as]: Save or remove?

File: Game/Resources/TestScene.json [hunk 4]
```diff
@@ -68,15 +68,97 @@
 				{
 					"Lens" : 
 					{
-						"fieldOfView" : 45.0,
+						"fieldOfView" : 45.000003814697266,
```
Comment (LarsIngo) [inline Game/Resources/TestScene.json]: Save or remove?

File: Game/Resources/TestScene.json [hunk 5]
```diff
@@ -328,7 +410,7 @@
 			{
 				"x" : 0.0,
 				"y" : 0.0,
-				"z" : 0.0
+				"z" : 400.0
```
Comment (LarsIngo) [inline Game/Resources/TestScene.json]: Shouldn't controls be relative payer? Set to 0?

File: src/Editor/main.cpp [hunk 6]
```diff
@@ -98,7 +98,7 @@ int main() {
 
                 { PROFILE("Render");
                 { GPUPROFILE("Render", Video::Query::Type::TIME_ELAPSED);
-                    Hymn().Render();
+                    Hymn().Render(nullptr, false, false, false, false, true, false);
```
Comment (LarsIngo) [inline src/Editor/main.cpp]: Why?

File: src/Editor/main.cpp.orig [hunk 7]
```diff
@@ -0,0 +1,152 @@
+#include <GL/glew.h>
```
Comment (LarsIngo) [inline src/Editor/main.cpp.orig]: Remove

File: src/Engine/Component/Controller.cpp [hunk 8]
```diff
@@ -0,0 +1,50 @@
+#include "Controller.hpp"
+#include "Manager/Managers.hpp"
+#include <Utility/Log.hpp> // TMP
```
Comment (LarsIngo) [inline src/Engine/Component/Controller.cpp]: Remove Log ;)

File: src/Engine/Component/Controller.cpp [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        glm::mat4 ctrlOrientation = glm::mat4(
+            glm::vec4(ctrlRight, 0.f),
+            glm::vec4(ctrlUp, 0.f),
+            glm::vec4(ctrlForward, 0.f),
+            glm::vec4(0.f, 0.f, 0.f, 1.f)
+        );
+
+        glm::vec3 localPosition = ctrlPosition * Managers().vrManager->GetScale();
+        glm::mat4 localTranslationMatrix = glm::translate(glm::mat4(), localPosition);
+        glm::mat4 globalTranslationMatix = entity->GetModelMatrix() * (ctrlOrientation * localTranslationMatrix * glm::mat4());
```
Comment (LarsIngo) [inline src/Engine/Component/Controller.cpp]: remove scale matrix

File: src/Engine/Component/Controller.hpp [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    public:
+        /// Constructor
+        Controller();
+
+        /// Destructor
+        ~Controller();            
+        
+        /// Save the component.
+        /**
+        * @return JSON value to be stored on disk.
```
Comment (LarsIngo) [inline src/Engine/Component/Controller.hpp]: Add space

File: src/Engine/Input/Input.hpp [hunk 11]
```diff
@@ -43,6 +46,13 @@ class Input {
          * @return Whether the button was activated this frame.
          */
         bool CheckButton(int index) const;
+
+        /// Check if a button was activated this frame
+        /**
+         * @param controller The controller to check.
+         * @retirm Whether the button was activated this frame.
```
Comment (LarsIngo) [inline src/Engine/Input/Input.hpp]: "return"

File: src/Engine/Manager/RenderManager.cpp [hunk 12]
```diff
@@ -99,7 +100,7 @@ void RenderManager::Render(World& world, Entity* camera) {
         }
 
         // Render hmd.
-        if (hmdRenderSurface != nullptr) {
+        if (hmdRenderSurface != nullptr && camera->name != "Editor Camera") {
```
Comment (LarsIngo) [inline src/Engine/Manager/RenderManager.cpp]: Ok for now, should be changed in the future

File: src/Engine/Manager/RenderManager.cpp [hunk 13]
```diff
@@ -251,10 +252,16 @@ void RenderManager::Render(World& world, const glm::mat4& translationMatrix, con
...
 
             if (mesh->geometry != nullptr && mesh->geometry->GetType() == Video::Geometry::Geometry3D::STATIC) {
-                Entity* entity = mesh->entity; 
+                Entity* entity = mesh->entity;
+                Controller* controller = entity->GetComponent<Controller>();
                 // If entity does not have material, it won't be rendered.
                 if (entity->GetComponent<Material>() != nullptr) {
-                    renderer->DepthRenderStaticMesh(mesh->geometry, viewMatrix, projectionMatrix, entity->GetModelMatrix());
+                    if (controller != nullptr && hmdRenderSurface != nullptr) {
+                        glm::mat4 ctrlModelMatrix = controller->HandleTransformation(entity);
```
Comment (LarsIngo) [inline src/Engine/Manager/RenderManager.cpp]: Will be changed in the future, Controller will be rendered as any entity!

File: src/Engine/Manager/RenderManager.cpp [hunk 14]
```diff
@@ -274,8 +281,14 @@ void RenderManager::Render(World& world, const glm::mat4& translationMatrix, con
             if (mesh->geometry != nullptr && mesh->geometry->GetType() == Video::Geometry::Geometry3D::STATIC) {
                 Entity* entity = mesh->entity;
                 Material* material = entity->GetComponent<Material>();
+                Controller* controller = entity->GetComponent<Controller>();
                 if (material != nullptr) {
-                    renderer->RenderStaticMesh(mesh->geometry, material->albedo->GetTexture(), material->normal->GetTexture(), material->metallic->GetTexture(), material->roughness->GetTexture(), entity->GetModelMatrix(), mesh->GetSelected());
+                    if (controller != nullptr && hmdRenderSurface != nullptr) {
+                        glm::mat4 ctrlModelMatrix = controller->HandleTransformation(entity);
```
Comment (LarsIngo) [inline src/Engine/Manager/RenderManager.cpp]: Will be changed in the future, Controller will be rendered as any entity!

File: Game/Resources/TestScene.json (hunk 5 above)
Comment (AncientFungus) [reply Game/Resources/TestScene.json]: This is cave's position. It shouldn't be 400 in z anyway but it has nothing to do with the player's position.

File: Game/Resources/Lever.as [hunk 15]
```diff
@@ -15,13 +15,11 @@ class Lever {
...
     void Update(float deltaTime){
         if (Input(MouseOpen)){
-            if(IsIntersect(self, player)){
-                if(self.rotation.y <= -45.0f){
-                    SendMessage(sibling, 1);
-                }
-                else
-                    self.rotation.y -= 0.3;
-            }
+			if(self.rotation.y <= -45.0f){
```
Comment (Chainsawkitten) [inline Game/Resources/Lever.as]: Tab

File: src/Engine/Component/Controller.hpp [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+        /**
+         * @return JSON value to be stored on disk.
+         */
+        Json::Value Save() const override;
+
+        /// Get's the VR controller's transformation matrix
+        /**
+         * @return The transformation matrix.
+         */
+        glm::mat4 HandleTransformation(Entity* entity);
```
Comment (Chainsawkitten) [inline src/Engine/Component/Controller.hpp]: Undocumented parameter

File: src/Engine/Component/Controller.hpp [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+        /// Handles all VR controller inputs (class for future implementations)
+        bool HandleInput(int buttonID);
```
Comment (Chainsawkitten) [inline src/Engine/Component/Controller.hpp]: Undocumented parameter and return value

File: src/Engine/Component/Controller.hpp [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+#include "../Entity/Entity.hpp"
+#include <glm/glm.hpp>
+#include <openvr.h>
+#include <glm/gtc/matrix_transform.hpp>
+
+namespace Component {
+    /// Component giving VR controller functions to an Entity
+    class Controller : public SuperComponent {
+
+    public:
```
Comment (Chainsawkitten) [inline src/Engine/Component/Controller.hpp]: Modifiers should be indented according to code standard.

File: src/Engine/Component/Mesh.hpp [hunk 19]
```diff
@@ -32,14 +32,14 @@ namespace Component {
 
             /// Get mesh selection.
             /**
-            * @return Bool Is this selected or not.
-            */
+             * @return Bool Is this selected or not.
```
Comment (Chainsawkitten) [inline src/Engine/Component/Mesh.hpp]: The type of the return value shouldn't be documented. Doxygen already extracts that.

File: src/Engine/Component/Mesh.hpp [hunk 20]
```diff
@@ -32,14 +32,14 @@ namespace Component {
...
+             */
             bool GetSelected() const;
 
             /// Set mesh selection.
             /**
-            * @param Value If selected or not.
-            */
+             * @param Value If selected or not.
```
Comment (Chainsawkitten) [inline src/Engine/Component/Mesh.hpp]: Parameter mismatch, upper-case V vs. lower-case.

File: src/Engine/Input/Input.hpp [hunk 21]
```diff
@@ -43,6 +46,13 @@ class Input {
          * @return Whether the button was activated this frame.
          */
         bool CheckButton(int index) const;
+
+        /// Check if a button was activated this frame
+        /**
+         * @param controller The controller to check.
+         * @return Whether the button was activated this frame.
+         */
+        bool CheckVRButton(int index, Component::Controller *controller) const;
```
Comment (Chainsawkitten) [inline src/Engine/Input/Input.hpp]: Undocumented parameter.

File: src/Engine/Manager/VRManager.hpp [hunk 22]
```diff
@@ -29,6 +30,12 @@ class VRManager {
          */
         glm::mat4 GetHMDPoseMatrix() const;
 
+        /// Returns the transform for the controllers
+        /**
+         * @return The Controllers translation matrix.
+         */
+        glm::mat4 GetControllerPoseMatrix(int controlID) const;
```
Comment (Chainsawkitten) [inline src/Engine/Manager/VRManager.hpp]: Undocumented parameter

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 10 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [RenderManager.cpp#L319](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/540687684fcfdb0dbf272c692aff14068b1b5a13/src/Engine/Manager/RenderManager.cpp#L319): Variable 'name' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [RenderManager.cpp#L329](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/540687684fcfdb0dbf272c692aff14068b1b5a13/src/Engine/Manager/RenderManager.cpp#L329): The function 'GetAnimations' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L347](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/540687684fcfdb0dbf272c692aff14068b1b5a13/src/Engine/Manager/RenderManager.cpp#L347): The function 'GetDirectionalLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L386](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/540687684fcfdb0dbf272c692aff14068b1b5a13/src/Engine/Manager/RenderManager.cpp#L386): The function 'GetMaterials' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L404](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/540687684fcfdb0dbf272c692aff14068b1b5a13/src/Engine/Manager/RenderManager.cpp#L404): The function 'GetMeshes' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L424](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/540687684fcfdb0dbf272c692aff14068b1b5a13/src/Engine/Manager/RenderManager.cpp#L424): The function 'GetPointLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L445](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/540687684fcfdb0dbf272c692aff14068b1b5a13/src/Engine/Manager/RenderManager.cpp#L445): The function 'GetSpotLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L549](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/540687684fcfdb0dbf272c692aff14068b1b5a13/src/Engine/Manager/ScriptManager.cpp#L549): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [VRManager.cpp#L128](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/540687684fcfdb0dbf272c692aff14068b1b5a13/src/Engine/Manager/VRManager.cpp#L128): The function 'SetScale' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 10 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [RenderManager.cpp#L319](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/fb87405751fccdf12520384887ba29c9dd1016b1/src/Engine/Manager/RenderManager.cpp#L319): Variable 'name' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [RenderManager.cpp#L329](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/fb87405751fccdf12520384887ba29c9dd1016b1/src/Engine/Manager/RenderManager.cpp#L329): The function 'GetAnimations' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L347](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/fb87405751fccdf12520384887ba29c9dd1016b1/src/Engine/Manager/RenderManager.cpp#L347): The function 'GetDirectionalLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L386](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/fb87405751fccdf12520384887ba29c9dd1016b1/src/Engine/Manager/RenderManager.cpp#L386): The function 'GetMaterials' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L404](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/fb87405751fccdf12520384887ba29c9dd1016b1/src/Engine/Manager/RenderManager.cpp#L404): The function 'GetMeshes' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L424](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/fb87405751fccdf12520384887ba29c9dd1016b1/src/Engine/Manager/RenderManager.cpp#L424): The function 'GetPointLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L445](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/fb87405751fccdf12520384887ba29c9dd1016b1/src/Engine/Manager/RenderManager.cpp#L445): The function 'GetSpotLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L549](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/fb87405751fccdf12520384887ba29c9dd1016b1/src/Engine/Manager/ScriptManager.cpp#L549): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [VRManager.cpp#L128](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/fb87405751fccdf12520384887ba29c9dd1016b1/src/Engine/Manager/VRManager.cpp#L128): The function 'SetScale' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**